### PR TITLE
Introduce the notion of hidden indices

### PR DESCRIPTION
**Pull request by s1monw**
With the setting `index.hidden` an index can be marked as hidden which
prevents the index from being resolved in index expressions or with `_all` / `*`.
Yet, if the index is access explicitly or the expression resolves to an alias pointing
to a hidden index the index is accessible. This is mainly a use case to prevent accidential
access of a metadata or cluster private index from an ordinary search.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java [hunk 1]
```diff
@@ -112,30 +116,40 @@ public boolean allowAliasesToMultipleIndices() {
     }
 
     public void writeIndicesOptions(StreamOutput out) throws IOException {
-        out.write(id);
+        out.writeInt(id);
     }
 
     public static IndicesOptions readIndicesOptions(StreamInput in) throws IOException {
         //if we read from a node that doesn't support the newly added flag (allowAliasesToMultipleIndices)
         //we just receive the old corresponding value with the new flag set to true (default)
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java]: may I ask you to remove this comment too? it's about bw comp which we have dropped a while ago

File: core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java [hunk 2]
```diff
@@ -159,13 +175,22 @@ public static IndicesOptions fromMap(Map<String, Object> map, IndicesOptions def
...
-                "ignore_unavailable".equals(name) || "ignoreUnavailable".equals(name) ||
-                "allow_no_indices".equals(name) || "allowNoIndices".equals(name);
+        switch (name) {
+            case "expand_wildcards":
+            case "expandWildcards":
+            case "ignore_unavailable":
+            case "ignoreUnavailable":
+            case "allow_no_indices":
+            case "allowNoIndices":
+            case "include_hidden":
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java]: I think we are missing includeHidden here

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 3]
```diff
@@ -252,6 +258,8 @@ private IndexMetaData(Index index, long version, State state, int numberOfShards
         this.indexCreatedVersion = indexCreatedVersion;
         this.indexUpgradedVersion = indexUpgradedVersion;
         this.minimumCompatibleLuceneVersion = minimumCompatibleLuceneVersion;
+        // by default we treat indices that start with '.' as hidden
+        this.hidden = INDEX_HIDDEN_SETTING.exists(settings) ? INDEX_HIDDEN_SETTING.get(settings) : index.getName().startsWith(".");
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: question: will we see the setting hidden:true amongst the index settings, or will it be implicit?

File: core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java [hunk 4]
```diff
@@ -352,6 +354,8 @@ public void testIndexOptionsNoExpandWildcards() {
 
             results = indexNameExpressionResolver.concreteIndices(context, Strings.EMPTY_ARRAY);
             assertEquals(0, results.length);
+
+
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java]: can you remove these? :)

File: docs/reference/docs/index_.asciidoc [hunk 5]
```diff
@@ -398,3 +398,15 @@ $ curl -XPUT 'http://localhost:9200/twitter/tweet/1?timeout=5m' -d '{
...
 --------------------------------------------------
+
+[float]
+[[hidden]]
+=== Hidden Indices
+
+An index can be hidden by setting `index.hidden` to `true` when then index is created.
+A hidden index will not be resolved from expression ie. the index `.scripts` is hidden
+by default from `5.0` on and won't be deleted if `_all` indices are deleted unless the indices
+option `include_hidden` is specified. Yet, if the index is specified explitily or via an alias
```
Comment (javanna) [inline docs/reference/docs/index_.asciidoc]: s/explitily/explicitly

File: rest-api-spec/src/main/resources/rest-api-spec/api/indices.delete.json [hunk 6]
```diff
@@ -20,6 +20,11 @@
         "master_timeout": {
           "type" : "time",
           "description" : "Specify timeout for connection to master"
+        },
+        "include_hidden": {
```
Comment (javanna) [inline rest-api-spec/src/main/resources/rest-api-spec/api/indices.delete.json]: shouldn't this flag be added to many more rest spec (e.g. search)? Wherever we have expand_wildcards and co. I thought

**Comment by javanna** [comment]
left a few comments, one other general question is why do we consider aliases different compared to wildcard expressions? Shall we consider hiding those indices even when expanding aliases?

File: core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java]: man I don't wanna introduce more broken settings here it's enough that we have the others ?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 3 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: implicit as you see here

File: rest-api-spec/src/main/resources/rest-api-spec/api/indices.delete.json (hunk 6 above)
Comment (s1monw) [inline rest-api-spec/src/main/resources/rest-api-spec/api/indices.delete.json]: we have lots of them without `expand_wildcards` where they should be so I didn't add them

File: core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java (hunk 2 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java]: I agree, then why did you add it above? :)

File: rest-api-spec/src/main/resources/rest-api-spec/api/indices.delete.json (hunk 6 above)
Comment (javanna) [inline rest-api-spec/src/main/resources/rest-api-spec/api/indices.delete.json]: ok good to know, maybe those are considered default params by the clients, not sure.

File: docs/reference/docs/index_.asciidoc [hunk 7]
```diff
@@ -398,3 +398,15 @@ $ curl -XPUT 'http://localhost:9200/twitter/tweet/1?timeout=5m' -d '{
...
+the index is not treated as hidden. This is useful for indices that should not be included if for instance
+`_all` is searched or if indices are currently re-indexed and follow some kind of index name pattern. This
+setting can be updated.
```
Comment (clintongormley) [inline docs/reference/docs/index_.asciidoc]: How about this:

    [float]
    [[hidden-indices]]
    === Hidden Indices

    added[5.0.0]

    An index can be hidden by setting `index.hidden` to `true`. This setting
    defaults to `false` unless the index name starts with `.`, such as `.kibana`.
    This is a dynamic setting which can be updated on an existing index.

    A hidden index will not be returned by multi-index expressions like `_all`, `*`,
    or `.marvel-*`. For instance, deleting `_all` indices will not delete hidden
    indices. Hidden indices can be referenced directly by name (eg `.kibana`) or
    by an alias that points to the index (eg `marvel-current`).

    Hidden indices can be revealed by including the `include_hidden` query string
    parameter, such as:

    [source,js]
    --------------------
    DELETE /_all?include_hidden
    --------------------

    or:

    [source,js]
    --------------------
    GET .marvel-*/_count?include_hidden
    --------------------

File: core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/support/IndicesOptions.java]: mistake - I hate those I think we should remove all of the came case ones

File: rest-api-spec/src/main/resources/rest-api-spec/api/indices.delete.json (hunk 6 above)
Comment (s1monw) [inline rest-api-spec/src/main/resources/rest-api-spec/api/indices.delete.json]: our test client fails :D

**Comment by s1monw** [comment]
> Shall we consider hiding those indices even when expanding aliases?

if you do that this stuff get's very hairy. I think aliases are different here they don't happen by accident they are like an explicit index. if you sym-link a hidden directory on disk it shows up too though

File: rest-api-spec/src/main/resources/rest-api-spec/api/indices.delete.json (hunk 6 above)
Comment (javanna) [inline rest-api-spec/src/main/resources/rest-api-spec/api/indices.delete.json]: oh.... :D

**Comment by javanna** [comment]
> if you sym-link a hidden directory on disk it shows up too though

I totally see this point, I think I was focusing mainly on aliases that point to multiple indices though... that break the directory analogy I believe. Fine with me though!

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 8]
```diff
@@ -301,6 +309,13 @@ public Version getUpgradedVersion() {
         return minimumCompatibleLuceneVersion;
     }
 
+    /**
+     * Retruns <code>true</code> if the index is a hidden index
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: Typo

File: rest-api-spec/src/main/resources/rest-api-spec/api/indices.delete.json (hunk 6 above)
Comment (clintongormley) [inline rest-api-spec/src/main/resources/rest-api-spec/api/indices.delete.json]: they're not considered default params - if they're missing from the rest specs, they're not supported in the clients

**Comment by s1monw** [comment]
@javanna @clintongormley I think it's ready

**Comment by clintongormley** [comment]
LGTM - we still need to add all the index options to the specs for all rest api's which support multi-indices, but i can do that in a follow up PR

**Comment by s1monw** [comment]
> LGTM - we still need to add all the index options to the specs for all rest api's which support multi-indices, but i can do that in a follow up PR

@clintongormley I think they are not added anywhere really

**Comment by dakrone** [comment]
@javanna I think this is waiting for your final +1 before being merged?

**Comment by javanna** [comment]
> @javanna I think this is waiting for your final +1 before being merged?

not really...but the PR was opened 6 months ago so who knows why it wasn't merged. I guess we decided not to do it? @s1monw do you remember?

**Comment by lukas-vlcek** [comment]
Will it be possible to learn which indices are hidden in `_cat` API?

**Comment by s1monw** [comment]
I am going to close this - we won't do this any time soon
